libi3/draw_util.c

```diff
@@ -28,6 +28,76 @@ static bool surface_initialized(surface_t *surface) {
     return true;
 }
 
+/*
+ * Get a GC for the given depth. The given drawable must have this depth.
+ *
+ * Per the X11 protocol manual for "CreateGC":
+ * > The gcontext can be used with any destination drawable having the same root
+ * > and depth as the specified drawable;
+ */
+static xcb_gcontext_t get_gc(xcb_connection_t *conn, uint8_t depth, xcb_drawable_t drawable, bool *should_free) {
+    static struct {
+        uint8_t depth;
+        xcb_gcontext_t gc;
+    } gc_cache[2] = {
+        0,
+    };
+
+    size_t index = 0;
+    bool cache = false;
+
+    *should_free = false;
+    for (; index < sizeof(gc_cache) / sizeof(gc_cache[0]); index++) {
+        if (gc_cache[index].depth == depth) {
+            return gc_cache[index].gc;
+        }
+        if (gc_cache[index].depth == 0) {
+            cache = true;
+            break;
+        }
+    }
+
+    xcb_gcontext_t gc = xcb_generate_id(conn);
+    xcb_void_cookie_t gc_cookie = xcb_create_gc_checked(conn, gc, drawable, 0, NULL);
```

Comment:
```suggestion
    /* The drawable is only used to get the root and depth, thus the gc is not
     * tied to the drawable and it can be re-used from different drawables. */
    xcb_void_cookie_t gc_cookie = xcb_create_gc_checked(conn, gc, drawable, 0, NULL);
```

I came up with this while trying to understand why this is safe to do e.g. when creating a gc from a drawable that is later freed. If my comment is correct, maybe we can include it for further context to readers?

Reply:
Done.

"Definitive answer" seems to come from the X11 protocol reference manual: https://www.x.org/releases/X11R7.6/doc/xproto/x11protocol.html#requests:CreateGC

>  The gcontext can be used with any destination drawable having the same root and depth as the specified drawable [...]